server/market/orderrouter.go

```diff
@@ -377,10 +441,97 @@ func (r *OrderRouter) handleMarket(user account.AccountID, msg *msgjson.Message)
 		req:   market,
 		msgID: msg.ID,
 	}
-	if err := tunnel.SubmitOrder(oRecord); err != nil {
-		log.Warnf("Market failed to SubmitOrder: %v", err)
-		return msgjson.NewError(msgjson.UnknownMarketError, "failed to submit order")
-	}
+
+	checkCoins := func() (tryAgain bool, msgErr *msgjson.Error) {
+		var valSum uint64
+		var spendSize uint32
+		for _, coin := range market.Trade.Coins {
+			// Get the coin from the backend and validate it.
+			dexCoin, err := fundingAsset.Backend.FundingCoin(coin.ID, coin.Redeem)
+			if err != nil {
+				if errors.Is(err, asset.CoinNotFoundError) {
+					return true, nil
+				}
+				return false, msgjson.NewError(msgjson.FundingError, fmt.Sprintf("error retrieving coin ID %v", coin.ID))
+			}
+
+			// Verify that the user controls the funding coins.
+			err = dexCoin.Auth(msgBytesToBytes(coin.PubKeys), msgBytesToBytes(coin.Sigs), coin.ID)
+			if err != nil {
+				log.Debugf("Auth error for %s coin %s: %v", fundingAsset.Symbol, dexCoin, err)
+				return false, msgjson.NewError(msgjson.CoinAuthError, fmt.Sprintf("failed to authorize coin %v", dexCoin))
+			}
+			valSum += dexCoin.Value()
+			spendSize += dexCoin.SpendSize()
+		}
+
+		if valSum == 0 {
+			return false, msgjson.NewError(msgjson.FundingError, "zero value funding coins not permitted")
+		}
+
+		// Calculate the fees and check that the utxo sum is enough.
+		var reqVal uint64
+		if sell {
+			lots := market.Quantity / assets.base.LotSize
+			reqVal = calc.RequiredOrderFunds(market.Quantity, uint64(spendSize), lots, &assets.funding.Asset)
+		} else {
+			// This is a market buy order, so the quantity gets special handling.
+			// 1. The quantity is in units of the quote asset.
+			// 2. The quantity has to satisfy the market buy buffer.
+			midGap := tunnel.MidGap()
+			buyBuffer := tunnel.MarketBuyBuffer()
+			lotWithBuffer := uint64(float64(assets.base.LotSize) * buyBuffer)
+			minReq := matcher.BaseToQuote(midGap, lotWithBuffer)
+			if minReq == 0 {
+				minReq = 1
+			}
```

Comment:
I actually didn't mean to leave this in here since there's already a quantity==0 rejection in `checkPrefixTrade`, but your suggestion is pretty reasonable so I'll go with the `midGap` check you suggested.